v3/lints/etsi/lint_qcstatem_etsi_national_scheme.go

```diff
@@ -0,0 +1,144 @@
+3) "PSD" for identification based on national authorization number of a payment service provider under
+Payments Services Directive (EU) 2015/2366 [i.13]. This shall use the extended structure as defined in ETSI
+TS 119 495 [3], clause 5.2.1.
+4) "LEI" for a global Legal Entity Identifier as specified in ISO 17442 [4]. The 2 character ISO 3166 [2] country
+code shall be set to 'XG'.
+5) Two characters according to local definition within the specified country and name registration authority,
+identifying a national scheme that is considered appropriate for national and European level, followed by the
+character ":" (colon)
+*************************************************************************/
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_qcstatem_correct_national_scheme",
+			Description:   "This lint checks that the national scheme is well-formed when used in the serialNumber or organizationIdentifier attribute in the subject field, provided that either the natural person semantics identifier or the legal person semantics identifier is present.",
+			Citation:      "ETSI EN 319 412-1 V1.4.1, 5.1.3 Natural person semantics identifier and 5.1.4 Legal person semantics identifier",
+			Source:        lint.EtsiEsi,
+			EffectiveDate: util.ETSI_EN_319_412_1_V1_4_1_DATE,
+		},
+		Lint: NewQcStatemNationalScheme,
+	})
+}
+
+func NewQcStatemNationalScheme() lint.LintInterface {
+	return &qcStatemNationalScheme{}
+}
+
+func (l *qcStatemNationalScheme) CheckApplies(c *x509.Certificate) bool {
+	_, isPresent := util.IsQcStatemPresent(c, &util.IdQcsPkixQCSyntaxV2)
+
+	if !isPresent {
+		return false
+	}
+
+	qcs2Generic := util.ParseQcStatem(util.GetQcStatemExtValue(c), util.IdQcsPkixQCSyntaxV2)
+
+	qcs2 := qcs2Generic.(util.DecodedQcS2)
+	semanticsId := qcs2.Decoded.SemanticsId
+	re := regexp.MustCompile(`^.{2}:`)
+
+	if semanticsId.Equal(util.IdEtsiQcsSemanticsIdNatural) {
+		serialNumber := c.Subject.SerialNumber
+		return re.MatchString(serialNumber)
+	}
+
+	if semanticsId.Equal(util.IdEtsiQcsSemanticsIdLegal) {
+		for _, orgId := range c.Subject.OrganizationIDs {
+			if re.MatchString(orgId) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (l *qcStatemNationalScheme) Execute(c *x509.Certificate) *lint.LintResult {
+
+	qcs2Generic := util.ParseQcStatem(util.GetQcStatemExtValue(c), util.IdQcsPkixQCSyntaxV2)
+
+	qcs2 := qcs2Generic.(util.DecodedQcS2)
+	semanticsId := qcs2.Decoded.SemanticsId
+
+	if semanticsId.Equal(util.IdEtsiQcsSemanticsIdNatural) {
+		serialNumber := c.Subject.SerialNumber
+		if !util.CheckNationalScheme(serialNumber) {
+			return &lint.LintResult{Status: lint.Error, Details: fmt.Sprintf("invalid format of subject:serialNumber %s for national scheme", serialNumber)}
+		}
+	}
+
+	if semanticsId.Equal(util.IdEtsiQcsSemanticsIdLegal) {
+		re := regexp.MustCompile(`^.{2}:`)
```

Comment:
Would you mind placing this regex in the global context so that `CheckApplies` and `Execute` use the precise same object?

Reply:
changed